core/os/os2/process_linux.odin

```diff
@@ -71,25 +299,439 @@ _Sys_Process_Attributes :: struct {}
+			cenv[i] = temp_cstring(env) or_return
+		}
+		env = &cenv[0]
+	}
+
+	child_pipe_fds: [2]linux.Fd
+	if errno = linux.pipe2(&child_pipe_fds, {.CLOEXEC}); errno != .NONE {
+		return process, _get_platform_error(errno)
+	}
+
+	// TODO: This is the traditional textbook implementation with fork.
+	//       A more efficient implementation with vfork:
+	//
+	//       1. retrieve signal handlers
+	//       2. block all signals
+	//       3. allocate some stack space
+	//       4. vfork (waits for child exit or execve); In child:
+	//           a. set child signal handlers
+	//           b. set up any necessary pipes
+	//           c. execve
+	//       5. restore signal handlers
+	//
+	pid: linux.Pid
+	if pid, errno = linux.fork(); errno != .NONE {
+		return process, _get_platform_error(errno)
+	}
+
+	STDIN  :: linux.Fd(0)
+	STDOUT :: linux.Fd(1)
+	STDERR :: linux.Fd(2)
+
+	READ :: 0
+	WRITE :: 1
+
+	if pid == 0 {
+		// in child process now
+		write_errno_to_parent_and_abort :: proc(parent_fd: linux.Fd, errno: linux.Errno) -> ! {
+			error_byte: [1]u8 = { u8(errno) }
+			linux.write(parent_fd, error_byte[:])
+			intrinsics.trap()
+		}
+
+		stdin_fd: linux.Fd
+		stdout_fd: linux.Fd
+		stderr_fd: linux.Fd
+
+		if desc.stdin != nil {
+			stdin_fd = linux.Fd(fd(desc.stdin))
+		} else {
+			stdin_fd, errno = linux.open("/dev/null", {})
+			if errno != .NONE {
+				write_errno_to_parent_and_abort(child_pipe_fds[WRITE], errno)
+			}
+		}
+
+		write_devnull: linux.Fd = -1
+
+		if desc.stdout != nil {
+			stdout_fd = linux.Fd(fd(desc.stdout))
+		} else {
+			write_devnull, errno = linux.open("/dev/null", {.WRONLY})
+			if errno != .NONE {
+				write_errno_to_parent_and_abort(child_pipe_fds[WRITE], errno)
+			}
+			stdout_fd = write_devnull
+		}
+
+		if desc.stderr != nil {
+			stderr_fd = linux.Fd(fd(desc.stderr))
+		} else {
+			if write_devnull == -1 {
+				write_devnull, errno = linux.open("/dev/null", {.WRONLY})
+				if errno != .NONE {
+					write_errno_to_parent_and_abort(child_pipe_fds[WRITE], errno)
+				}
+			}
+			stderr_fd = write_devnull
+		}
+
+		if _, errno = linux.dup2(stdin_fd, STDIN); errno != .NONE {
+			write_errno_to_parent_and_abort(child_pipe_fds[WRITE], errno)
+		}
+		if _, errno = linux.dup2(stdout_fd, STDOUT); errno != .NONE {
+			write_errno_to_parent_and_abort(child_pipe_fds[WRITE], errno)
+		}
+		if _, errno = linux.dup2(stderr_fd, STDERR); errno != .NONE {
+			write_errno_to_parent_and_abort(child_pipe_fds[WRITE], errno)
+		}
+
+		success_byte: [1]u8
+		linux.write(child_pipe_fds[WRITE], success_byte[:])
+
+		if errno = linux.execveat(exe_fd, "", &cargs[0], env, {.AT_EMPTY_PATH}); errno != .NONE {
+			write_errno_to_parent_and_abort(child_pipe_fds[WRITE], errno)
+		}
+		unreachable()
+	}
+
+	linux.close(child_pipe_fds[WRITE])
+	defer linux.close(child_pipe_fds[WRITE])
+
+	n: int
+	child_byte: [1]u8
+	n, errno = linux.read(child_pipe_fds[READ], child_byte[:])
+	if errno != .NONE {
+		return process, _get_platform_error(errno)
```

Comment:
Can't this throw BROKEN_PIPE?

I thought that if the child is started successfully, the pipe will be closed and errno set to broken pipe.

Reply:
I don't think it can without it being terminated outside of our control.  Just to verify, I added a 1 second busy loop right before this read and ran it through a set of tests that mostly run programs that complete near instantly and it still reads just fine.  The thinking here was that everything is still valid until the process is reaped.  Now... for the life of me.. I cannot find documentation that says that anywhere.

But, I'm glad you pointed this out.  The error handling doesn't wait on the process and leaks the PID.

Reply:
Check out the update then.. I just stopped trying to handle the read error all together.  If the read fails, something extraordinary happened, but the user now has a process to wait on.  They can figure it out on that end.